layabout.py

```diff
@@ -1,3 +1,287 @@
+        """
+        resetting = False
+
+        # If we were given a token let's prefer the new one and establish or
+        # reset the connection.
+        if token is not None:
+            self._token = token
+            resetting = True
+
+        # We don't have an existing token, so let's try to get one from an
+        # environment variable or give up.
+        if self._token is None:
+            try:
+                self._token = os.environ[self._env_var]
+            except KeyError:
+                raise MissingSlackToken('Cannot connect to the Slack API'
+                                        ' without a token')
+
+        # Either we've never connected before or we're purposefully resetting
+        # the connection.
+        if self.slack is None or resetting:
+            # TODO: Maybe set an appropriate user agent string here using
+            # self.name.
+            self.slack = SlackClient(token=self._token)
+
+        # Use whatever token we've got to attempt to connect (or reconnect).
+        return self.slack.rtm_connect()
+
+    def _reconnect(self, retries: int,
+                   backoff: Callable[[int], float]) -> bool:
+        """
+        Attempt to reconnect to the Slack API.
+
+        Args:
+            retries: The number of retry attempts to make if a connection
+                to Slack if not established or is lost.
+            backoff: The strategy used to determine how
+                long to wait between retries. Must take as input the number of
+                the current retry and output a :obj:`float`.
+
+        Returns:
+            Whether the reconnection succeeded.
+        """
+        # TODO: Should retries start at 0 or 1?
+        for retry in range(retries):
+            if self.connect():
+                return True
+            else:
+                interval = backoff(retry)
+                log.debug("Waiting %.3fs before retrying", interval)
+                time.sleep(interval)
+
+        return False
+
+    def run(self, token: str = None, interval: float = 0.5,
+            retries: int = 4, backoff: Callable[[int], float] = None,
+            until: Callable[[List[dict]], bool] = None) -> None:
+        """
+        Connect to the Slack API and run the event handler loop.
+
+        Args:
+            token: A Slack API token. If absent an attempt will be made to use
+                the environment variable supplied at instantiation.
+            interval: The number of seconds to wait between fetching events
+                from the Slack API.
+            retries: The number of retry attempts to make if a connection to
+                Slack if not established or is lost.
+            backoff: The strategy used to determine how long to wait between
+                retries. Must take as input the number of the current retry and
+                output a :obj:`float`. If absent an exponential backoff
+                strategy will be used.
```

Comment:
I'm used to most modern application backoff strategies using a random cumulative backoff like the [Google Cloud Storage one](https://cloud.google.com/storage/docs/exponential-backoff). Not sure how much that matters for you. It just avoids a class of synchronization bugs.

Reply:
Huh. I'm not familiar with that. I'll have to go read up on that. Thanks. :smile:

Reply:
Turned out to be a great suggestion. I've incorporated this now. Thanks! :grin: